### PR TITLE
fix(db): prevent SQLITE_NOTADB corruption under concurrent sub-agents

### LINKED ISSUE #2682: Prevent SQLite corruption (`SQLITE_NOTADB`/code 26) under concurrent sub-agents
### Problem

Crush can fail to start with:

```
Failed to connect to database: file is not a database (26).
```

This occurred after a session that ran many sub-agents in parallel (e.g. the LLM invoking \~10 `agent` tool calls concurrently via `fantasy.NewParallelAgentTool`). The `.crush/crush.db` file ends up in a state where SQLite no longer recognizes its header, making the project session unrecoverable without deleting the DB.

I had Opus 4.7 investigate what could cause this, and here's what it came up with:

> Root cause: in `internal/db/connect.go` the `*sql.DB` is opened with Go's default pool settings — unlimited `MaxOpenConns`, no `_txlock`. The same `*sql.DB` is shared by `session`, `message`, `history`, and `filetracker` services (`internal/app/app.go`), and every sub-agent performs session/message/cost writes of its own (`coordinator.runSubAgent`, `coordinator.go:1014`). Under N parallel sub-agents:
>
> 1. The pool opens multiple concurrent SQLite handles against the same file.
> 2. WAL frames from different handles interleave; each handle can trigger auto-checkpoints.
> 3. If the process is cancelled/killed mid-checkpoint (context cancellation of parent agents, SIGINT, OOM), the main DB file header and WAL can desync.
> 4. Next open → `SQLITE_NOTADB (26)`.
>
> Deferred transactions also allow a classic upgrade-deadlock pattern (two readers both trying to `BEGIN` → write) which contributed to previously-reported `SQLITE_BUSY` issues (see [CRUSHGH-678](https://linear.app/charmland/issue/CRUSHGH-678/sqlite-busy-dead-lock), [CRUSHGH-815](https://linear.app/charmland/issue/CRUSHGH-815/crush-0551-archaur-yay-crush-bin-erroring-with-sqlite-busy)) — related but distinct from NOTADB.

### Solution

Here's the solution Opus 4.7 came up with:

Two small changes to the DB layer:

1. `db.SetMaxOpenConns(1)` — SQLite serializes writes at the file level regardless of pool size, so multiple connections add no throughput, only race surface. Forcing a single pooled connection makes Go's `database/sql` queue all callers through one SQLite handle, eliminating WAL/checkpoint races across handles. Sub-agents still work; they just serialize their writes, which SQLite was going to do anyway.
2. `_txlock=immediate` in the DSN for both drivers (`modernc.org/sqlite` and `ncruces/go-sqlite3`) — makes `BEGIN` acquire the reserved-writer lock up front, so concurrent writers queue cleanly on `busy_timeout` (30s) rather than racing the deferred→write upgrade. Defense-in-depth alongside the pool limit.

For the ncruces driver the DSN must use the `file:` URI prefix for query params to be parsed (confirmed in `ncruces/go-sqlite3 driver.newConnector`); the plain path form silently ignores `_txlock`.

<details><summary>Patch</summary>

```diff
diff --git a/internal/db/connect.go b/internal/db/connect.go
index 5cff191c..7ba8bdc7 100644
--- a/internal/db/connect.go
+++ b/internal/db/connect.go
@@ -45,6 +45,13 @@ func Connect(ctx context.Context, dataDir string) (*sql.DB, error) {
 		return nil, err
 	}
 
+	// Serialize all access through a single connection. SQLite serializes
+	// writes at the file level anyway, and allowing multiple pool
+	// connections to interleave writes/checkpoints (especially under
+	// concurrent sub-agents) has caused WAL/header desync resulting in
+	// SQLITE_NOTADB (26) on the next open.
+	db.SetMaxOpenConns(1)
+
 	if err = db.PingContext(ctx); err != nil {
 		db.Close()
 		return nil, fmt.Errorf("failed to connect to database: %w", err)
diff --git a/internal/db/connect_modernc.go b/internal/db/connect_modernc.go
index 39c7faa4..2e9676d6 100644
--- a/internal/db/connect_modernc.go
+++ b/internal/db/connect_modernc.go
@@ -17,6 +17,9 @@ func openDB(dbPath string) (*sql.DB, error) {
 	for name, value := range pragmas {
 		params.Add("_pragma", fmt.Sprintf("%s(%s)", name, value))
 	}
+	// Use BEGIN IMMEDIATE so writers acquire the reserved lock up front,
+	// preventing deferred-to-writer upgrade deadlocks.
+	params.Set("_txlock", "immediate")
 
 	dsn := fmt.Sprintf("file:%s?%s", dbPath, params.Encode())
 	db, err := sql.Open("sqlite", dsn)
diff --git a/internal/db/connect_ncruces.go b/internal/db/connect_ncruces.go
index 12946e29..d9083335 100644
--- a/internal/db/connect_ncruces.go
+++ b/internal/db/connect_ncruces.go
@@ -11,7 +11,11 @@ import (
 )
 
 func openDB(dbPath string) (*sql.DB, error) {
-	db, err := driver.Open(dbPath, func(c *sqlite3.Conn) error {
+	// Use BEGIN IMMEDIATE so writers acquire the reserved lock up front,
+	// preventing deferred-to-writer upgrade deadlocks. The "file:" prefix
+	// is required for the ncruces driver to parse query parameters.
+	dsn := fmt.Sprintf("file:%s?_txlock=immediate", dbPath)
+	db, err := driver.Open(dsn, func(c *sqlite3.Conn) error {
 		// Set pragmas for better performance via _pragma query params.
 		// Format: PRAGMA name = value;
 		for name, value := range pragmas {
```

+++

</details>

These changes have been implemented in [my fork of crush](<https://github.com/taoeffect/crush>) (which has [these other features](<https://github.com/charmbracelet/crush/pulls/taoeffect>)).

### Version

v0.60.x

### Environment

Fedora

### PR DESCRIPTION
**Pull request by taoeffect**
Closes #2682

This fix is authored by Opus 4.7 using crush. See #2682 for details.

Crush could fail to start with "file is not a database (26)" after a session ran many sub-agents in parallel. The shared *sql.DB was opened with Go's default unlimited connection pool, so each concurrent sub-agent (session/message/cost writes via coordinator.runSubAgent) could open its own SQLite handle. Under load, WAL frames from different handles interleaved and auto-checkpoints could race; if the process was cancelled or killed mid-checkpoint, the main DB header and WAL desynced, producing SQLITE_NOTADB (26) on the next open.

Two changes address this:

- Set db.SetMaxOpenConns(1). SQLite serializes writes at the file level anyway, so extra pool connections add no throughput, only race surface. Forcing a single pooled connection makes database/sql queue all callers through one SQLite handle, eliminating cross-handle WAL/checkpoint races. Sub-agents still function; their writes just serialize (which SQLite was going to do regardless).

- Add _txlock=immediate to both driver DSNs (modernc.org/sqlite and ncruces/go-sqlite3). BEGIN IMMEDIATE acquires the reserved-writer lock up front so concurrent writers queue cleanly on busy_timeout instead of racing the deferred-to-writer upgrade (the pattern behind prior SQLITE_BUSY reports, e.g. #2129, #2576). Defense in depth alongside the pool limit.

For the ncruces driver the DSN must use the file: URI prefix for query params to be parsed; the plain path form silently ignores _txlock.

- [x] I have read [`CONTRIBUTING.md`](https://github.com/charmbracelet/.github/blob/main/CONTRIBUTING.md).
- [x] I have created a discussion that was approved by a maintainer (for new features).

### CODE REVIEW

File: internal/db/connect_ncruces.go [hunk 1]
```diff
@@ -11,7 +11,11 @@ import (
 )
 
 func openDB(dbPath string) (*sql.DB, error) {
-	db, err := driver.Open(dbPath, func(c *sqlite3.Conn) error {
+	// Use BEGIN IMMEDIATE so writers acquire the reserved lock up front,
+	// preventing deferred-to-writer upgrade deadlocks. The "file:" prefix
+	// is required for the ncruces driver to parse query parameters.
+	dsn := fmt.Sprintf("file:%s?_txlock=immediate", dbPath)
+	db, err := driver.Open(dsn, func(c *sqlite3.Conn) error {
```
[the remark this answers is not in the document]
Comment (taoeffect) [reply internal/db/connect_ncruces.go:18]: Had Opus 4.7 look into this, here's what it found:

-------------

I just tested empirically against both drivers with edge-case paths:

| Path contains | ncruces | modernc | Notes |
|---|---|---|---|
| space (`/tmp/sqlite test dir/`) | ✅ works | ✅ works | common case — no problem |
| `%` (`/tmp/sqlitePCT%dir/`) | ✅ works | ✅ works | neither driver percent-decodes the path |
| `#` (`/tmp/sqliteHASH#dir/`) | ❌ truncated | ❌ truncated | db created at `/tmp/sqliteHASH` |
| `?` (`/tmp/sqliteQ?dir/`) | ❌ truncated | ❌ truncated | db created at `/tmp/sqliteQ` |

Why the test result differs from the Copilot theory:

- ncruces only calls `url.ParseQuery` on the substring **after the first `?`**, and passes `n.name` (the whole DSN) straight to `sqlite3.OpenContext`. SQLite's URI parser is lenient about spaces and unencoded `%` in the path.
- So the realistic worry (spaces on Windows paths like `C:\Users\John Doe\...`) is a **non-issue** — it just works.
- Only `?` and `#` literally in the filesystem path actually break. `?` is forbidden on Windows filesystems, and `#` in a data-dir path is vanishingly rare.

So: **Copilot is technically right that the DSN isn't RFC-3986-safe, but the only characters that actually cause the described breakage (`?`, `#`) are essentially unreachable in real `CRUSH_GLOBAL_DATA` / `%LOCALAPPDATA%` / `$XDG_DATA_HOME` paths.** Spaces, `%`, and other common "reserved" chars work in practice.

Verdict: **safe to ignore** for this fix. If you want belt-and-braces, a one-line change to use `url.URL{Opaque: "/"+filepath.ToSlash(dbPath)}.String()` would close the remaining `?`/`#` hole, but it's not blocking the 2682 fix.

---------------------

I think it's fine to leave as-is but can also ask crush/opus to make this change, your call.

**Review by andreynering — APPROVED** [review]
Thank you @taoeffect!
